### PR TITLE
Windows build

### LINKED ISSUE #778: Fix Windows builds
**Describe the bug**

It was reported in https://github.com/skycoin/skywire/pull/772 that the Windows builds are broken. Skywire should build on Windows if following the setup steps [here](https://github.com/skycoin/skywire/blob/master/docs/windows-setup.md). 

We need to investigate the architecture of 

- `syslog`
- `dmsgpty` packages as well. 

They currently do not build but at least for `dmsgpty` we can likely get it to build.

### PR DESCRIPTION
**Pull request by alexadhy**
Did you run `make format && make check`? yes

Fixes #778 

 Changes:	
- Various fixes to make skywire build on windows

TODO:

- [x] `cmdutil` function such as `cmdutil.SignalContext` should probably be fixed on `dmsg` repo
- [x] many things are redundant (see `stdio` package for example) in this draft PR, probably refactor it later once it is tested

How to test this PR:
- Follow https://github.com/skycoin/dmsg/pull/83#issue-655883397
- See skywire wiki for running it on windows

NEEDS
- https://github.com/skycoin/dmsg/pull/93

### CODE REVIEW

**Review by jdknives — COMMENTED** [review]
Good start

**Comment by 4rchim3d3s** [comment]
For additional info on my comment on issue #778 

Also tried to build this PR on windows and got this output:

image

**Comment by alexadhy** [comment]
Screen Shot 2021-05-31 at 21 50 54


@4rchim3d3s : the build instruction is here => https://github.com/skycoin/dmsg/pull/83#issue-655883397

**Comment by 4rchim3d3s** [comment]
Thanks, I followed the instructions and instead of `make` I tried the `go build...`

And now got this:

`build github.com/skycoin/skywire/cmd/skywire-visor: cannot load embed: malformed module path "embed": missing dot in first path element`

**Comment by alexadhy** [comment]
@4rchim3d3s : I cannot reproduce.

Please follow the guide:

- Clone and checkout that PR (https://github.com/skycoin/dmsg/pulls/83) in your local dir (i.e. `C:\Users\user\Desktop\dmsg`
- Clone and checkout this PR to the adjacent dir `C:\Users\user\Desktop\skywire`
- Run this command **inside** skywire dir `go mod edit -replace github.com/skycoin/dmsg=../dmsg` 
- Then sync the vendor directory `go mod vendor -v`
- Then do `go build -o .\skywire-visor.exe .\cmd\skywire-visor\`

**Comment by 4rchim3d3s** [comment]
@alexadhy I retried it on desktop.

`go mod vendor -v` gave me: 

 ```sh
$ go mod vendor -v
go: downloading github.com/songgao/water v0.0.0-20200317203138-2b4b6d7c09d8
go: downloading github.com/syndtr/gocapability v0.0.0-20200815063812-42c35b437635
go: extracting github.com/syndtr/gocapability v0.0.0-20200815063812-42c35b437635
go: extracting github.com/songgao/water v0.0.0-20200317203138-2b4b6d7c09d8
go: downloading github.com/creack/pty v1.1.11
go: extracting github.com/creack/pty v1.1.11
go: downloading github.com/mmcloughlin/avo v0.0.0-20200523190732-4439b6b2c061
go: extracting github.com/mmcloughlin/avo v0.0.0-20200523190732-4439b6b2c061
go: downloading golang.org/x/tools v0.1.0
go: extracting golang.org/x/tools v0.1.0
go: downloading golang.org/x/xerrors v0.0.0-20200804184101-5ec99f83aff1
go: downloading golang.org/x/mod v0.4.2
go: extracting golang.org/x/xerrors v0.0.0-20200804184101-5ec99f83aff1
go: extracting golang.org/x/mod v0.4.2
github.com/skycoin/skywire/cmd/apps/skychat imports
        embed: malformed module path "embed": missing dot in first path element
github.com/skycoin/skywire/cmd/apps/skychat imports
        io/fs: malformed module path "io/fs": missing dot in first path element
```

And then with `go build...` 

```sh
$ go build -o ./skywire-visor.exe ./cmd/skywire-visor/
build github.com/skycoin/skywire/cmd/skywire-visor: cannot load embed: malformed module path "embed": missing dot in first path element
```

Guess I will just wait until it works as I don't know golang and how to import modules :) 

If you wish to I can try again someday but I don't want to make trouble because of my impatience :)

**Comment by alexadhy** [comment]
@4rchim3d3s I think you need to upgrade your golang version, you need go version 1.16.x to make it work

**Comment by 4rchim3d3s** [comment]
@alexadhy I am sorry. That was the problem. Now skywire and vpn app build without problem

File: pkg/restart/restart_windows.go [hunk 1]
```diff
@@ -2,10 +2,13 @@
...
+)
 
-func attachTTY(*exec.Cmd) {
+func attachTTY(_ *exec.Cmd) {
 	// not used for Windows
+}
 
-	return
+func (c *Context) ignoreSignals() {
 }
```
Comment (mrpalide) [inline pkg/restart/restart_windows.go:15]: What are these blank functions?

Comment (alexadhy) [reply pkg/restart/restart_windows.go:15]: placeholder function, to match the unix equivalent

**Comment by alexadhy** [comment]
What previous commits did:

- Adds systray support for windows 

Screen Shot 2021-09-10 at 15 51 29

tray:


Screen Shot 2021-09-10 at 15 58 35

File: Makefile [hunk 2]
```diff
@@ -19,7 +19,14 @@ BRANCH := latest
 
 PROJECT_BASE := github.com/skycoin/skywire
 DMSG_BASE := github.com/skycoin/dmsg
-OPTS?=GO111MODULE=on
+ifeq ($(OS),Windows_NT)
+	OPTS?=powershell -Command setx GO111MODULE on;
+	DATE := $(shell powershell -Command date -u ${RFC_3339})
+
+else
+	OPTS?=GO111MODULE=on
```
Comment (ersonp) [inline Makefile:24]: We need to use `DATE := $(shell date -u $(RFC_3339))` inside the else.

**Comment by ersonp** [comment]
In makefile we need to use
```make
ifeq ($(OS),Windows_NT)
	SHELL := pwsh
else
	SHELL := /bin/bash
endif
```
instead of this [here](https://github.com/alexadhy/skywire/blob/470cccd61f7ad082c2cc14d59a0d9a81c56c0c44/Makefile#L11)
```make
SHELL := /bin/bash
```

Also, we should create `make help` for windows too.

**Comment by ersonp** [comment]
We need to generate a `sock` file at ` C:\Users\<YOUR_USER>\dmsgpty.sock`

**Comment by ersonp** [comment]
```shell
[2021-09-14T12:22:08+05:30] INFO [visor:config]: Reading config from file. filepath="skywire-config.json"
[2021-09-14T12:22:08+05:30] DEBUG []: Process info delay=0s parent_systemd=false skybian_build_version="" systemd=false
Version "v0.5.0-rc1-35-g470cccd6" built on "2021-09-14T12:21:40Z" against commit "470cccd61f7ad082c2cc14d59a0d9a81c56c0c44"
[2021-09-14T12:22:08+05:30] INFO [visor:config]: Reading config from file. filepath="skywire-config.json"
[2021-09-14T12:22:08+05:30] INFO [visor:config]: Flushing config to file. config_version="v1.1.0" filepath="skywire-config.json"
[2021-09-14T12:22:08+05:30] INFO [visor:config]: Flushing config to file. config_version="v1.1.0" filepath="skywire-config.json"
[2021-09-14T12:22:08+05:30] INFO [visor:startup]: Begin startup. public_key=0358582eb1f67bde9f6f7209738b58eca83433a0a794f7837c7c0ca42ed0684607
[2021-09-14T12:22:08+05:30] INFO [hypervisor]: Starting
[2021-09-14T12:22:08+05:30] INFO [visor]: Starting
[2021-09-14T12:22:08+05:30] INFO [stcpr]: Starting
[2021-09-14T12:22:08+05:30] INFO [launcher]: Starting
[2021-09-14T12:22:08+05:30] INFO [updater]: Starting
[2021-09-14T12:22:08+05:30] INFO [dmsg]: Starting
[2021-09-14T12:22:08+05:30] INFO [updater]: Initialized in 0s (0s with dependencies)
[2021-09-14T12:22:08+05:30] INFO [event_broadcaster]: Starting
[2021-09-14T12:22:08+05:30] INFO [event_broadcaster]: Initialized in 0s (0s with dependencies)
[2021-09-14T12:22:08+05:30] INFO [address_resolver]: Starting
[2021-09-14T12:22:08+05:30] INFO [discovery]: Starting
[2021-09-14T12:22:08+05:30] INFO [dmsg_pty]: Starting
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x0 pc=0xd13032]

goroutine 21 [running]:
github.com/getlantern/systray.(*winTray).setTooltip(0x1a2be80, {0xe8def2, 0x0})
        C:/Users/Erson/skywire/vendor/github.com/getlantern/systray/systray_windows.go:238 +0xb2
github.com/getlantern/systray.SetTooltip({0xe8def2, 0xc0001c7328})
        C:/Users/Erson/skywire/vendor/github.com/getlantern/systray/systray_windows.go:899 +0x2f
github.com/skycoin/skywire/internal/gui.GetOnGUIReady.func1()
        C:/Users/Erson/skywire/internal/gui/gui.go:57 +0x56
github.com/getlantern/systray.Register.func2()
        C:/Users/Erson/skywire/vendor/github.com/getlantern/systray/systray.go:94 +0x32
created by github.com/getlantern/systray.Register
        C:/Users/Erson/skywire/vendor/github.com/getlantern/systray/systray.go:92 +0xc8
```

Normal skywire is working fine but I am getting this for systray. I'm guessing this has something to do with the icon but I do not 
 know the path to put it at.

**Comment by ersonp** [comment]
The shutdown logic is not working on skywire.

File: Makefile (hunk 2 above)
Comment (alexadhy) [reply Makefile:24]: there's already a DATE variable defined above, so no need to define it twice

Comment (ersonp) [reply Makefile:24]: Yes, what I meant was that `DATE := $(shell date -u $(RFC_3339))` should be inside the else statement as `make` isn't working in windows if it is outside.

**Comment by ersonp** [comment]
- Using `Ctrl+C` to close the visor closes it directly without properly shutting down the visor. It is working via `systray>quit`.
- Makefile is not working because of the above-mentioned `DATE` variable being outside the else statement.

Everything else has been fixed and is working well.

**Comment by alexadhy** [comment]
> - Using `Ctrl+C` to close the visor closes it directly without properly shutting down the visor. It is working via `systray>quit`.
> 
> - Makefile is not working because of the above-mentioned `DATE` variable being outside the else statement.
> 
> 
> 
> Everything else has been fixed and is working well.

Ah, I forgot to add the first point.

**Comment by ersonp** [comment]
- The shutdown logic is working now in powershell with `Ctrl+C` but `Ctrl+C` doesn't work in CMD. We should just mention that somewhere in README to only run Visor on `powershell` if it already isn't.
- Makefile is also working but I had a small suggestion. We should move `.DEFAULT_GOAL := help` inside the `ifeq ($(OS),Windows_NT)` as `.DEFAULT_GOAL := help-windows` and `.DEFAULT_GOAL := help` inside the `else` statement.

File: Makefile [hunk 3]
```diff
@@ -70,17 +82,27 @@ clean: ## Clean project: remove created binaries and apps
...
 
 lint: ## Run linters. Use make install-linters first
 	${OPTS} golangci-lint run -c .golangci.yml ./...
 	# The govet version in golangci-lint is out of date and has spurious warnings, run it separately
 
-lint-windows-appveyor:
+lint-windows: ## Run linters. Use make install-linters-windows first
+	powershell 'golangci-lint run -c .golangci.yml ./...'
+
+lint-appveyor-windows: ## Run linters for appveyor only on windows
```
Comment (ersonp) [inline Makefile:100]: This is giving an error
```
PS C:\Users\Erson\skywire> make lint-appveyor-windows
C:\Users\appveyor\go\bin\golangci-lint run -c .golangci.yml ./...
process_begin: CreateProcess(NULL, C:\Users\appveyor\go\bin\golangci-lint run -c .golangci.yml ./..., ...) failed.
make (e=2): The system cannot find the file specified.
make: *** [Makefile:106: lint-appveyor-windows] Error 2
```

**Comment by ersonp** [comment]
`build-windows-systray` is not showing up in `make help-windows`

File: Makefile (hunk 3 above)
Comment (alexadhy) [reply Makefile:100]: Have you tried to add `golangci-lint` to your own `%PATH%`, because that one is for appveyor

Comment (alexadhy) [reply Makefile:100]: and you should've run `make check-windows`

Comment (ersonp) [reply Makefile:100]: Okay. Was just testing all the commands. Everything else works then.

**Comment by alexadhy** [comment]
> `build-windows-systray` is not showing up in `make help-windows`

fixed

**Review by ersonp — APPROVED** [review]
Just add that this pr needs https://github.com/skycoin/dmsg/pull/113. Everything else looks good and is working.

**Comment by alexadhy** [comment]
This PR requires https://github.com/skycoin/dmsg/pull/113 to be merged first

**Comment by ersonp** [comment]
As mentioned [here](https://github.com/skycoin/dmsg/pull/108#issuecomment-922689723), `make lint-windows` gives a similar error after fixing the PATH warning.
[make lint-windows.txt](https://github.com/skycoin/skywire/files/7194150/make.lint-windows.txt)

**Comment by alexadhy** [comment]
> As mentioned [here](https://github.com/skycoin/dmsg/pull/108#issuecomment-922689723), `make lint-windows` gives a similar error after fixing the PATH warning.
> [make lint-windows.txt](https://github.com/skycoin/skywire/files/7194150/make.lint-windows.txt)


All you need to do is to run make format-windows. 

Those logs contain warn / err message for golangci-lint. Better yet, I’ll do it for you and push it.

**Comment by alexadhy** [comment]
### TODOs

- [ ] Create Installer with `wix` (In progress)
- [x] Update the wireguard-go version in windows.
- [x] Update `wintun` and its versions (consider also shipping wintun.dll with the installer)
- [x] Would be great if we eliminate the need to run the `VPN` as local system user
- [x] Fixes stopping application (hackish)

**Comment by alexadhy** [comment]
New commits add support for running vpn-client without `local system` privilege:

Instructions are in: https://github.com/skycoin/skywire/blob/74d9e9307400ee511a9147e52fc65e7f62d7f9c5/docs/windows-setup.md

Screen Shot 2021-10-13 at 09 51 28

**Comment by mrpalide** [comment]
I test it on test env. One point I see that was wrong for me:
- Full Terminal of remote visor not available due this error:
```
[dmsgpty-ui] Error: exec: "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe": executab
le file not found in $PATH                                                                          
[dmsgpty-ui] Status: Stopped!                                                                       
```

Full log: [log.txt](https://github.com/skycoin/skywire/files/7417566/log.txt)

**Comment by ersonp** [comment]
Getting nil pointer dereference in vpn-client when trying to connect to `033d5c3463cf5cba2bcdb0ff84e755c4e5cf63797ec5c93c165f0c02078bafcfa4` with an active dmsg transport to the same visor on prod.
[nilpointer.txt](https://github.com/skycoin/skywire/files/7543586/nilpointer.txt)

**Comment by ersonp** [comment]
> Getting nil pointer dereference in vpn-client when trying to connect to `033d5c3463cf5cba2bcdb0ff84e755c4e5cf63797ec5c93c165f0c02078bafcfa4` with an active dmsg transport to the same visor on prod. [nilpointer.txt](https://github.com/skycoin/skywire/files/7543586/nilpointer.txt)

This was because `wintun` was not setup. Vpn is working as intended.

File: Makefile [hunk 4]
```diff
@@ -8,17 +7,29 @@
 .PHONY : docker-run docker-stop
 
 SHELL := /bin/bash
-VERSION ?= $(shell git describe 2> /dev/null)
-#VERSION := v0.1.0 # for debugging updater
+
+VERSION := $(shell git describe)
 
 RFC_3339 := "+%Y-%m-%dT%H:%M:%SZ"
 DATE := $(shell date -u $(RFC_3339))
```
Comment (ersonp) [inline Makefile]: This needs to be removed. It's preventing `make` from being run on windows. We have a `DATE` var in the if statement below.

File: Makefile [hunk 5]
```diff
@@ -8,17 +7,29 @@
 .PHONY : docker-run docker-stop
 
 SHELL := /bin/bash
```
Comment (ersonp) [inline Makefile]: No need to have `SHELL` here as it is also in the if statement below.

**Comment by ersonp** [comment]
image

Everything is working fine now except the ping in the VPN. But I think this should be fixed in the VPN statistics PR. So this is good to go.

**Review by libronlopes — APPROVED** [review]

**Pull request merged**